internal/db/subscriptions.go

```diff
@@ -19,83 +19,75 @@ import (
 	"github.com/sourcenetwork/defradb/internal/planner"
 )
 
-func (db *db) checkForClientSubscriptions(r *request.Request) (
-	*events.Publisher[events.Update],
-	*request.ObjectSubscription,
-	error,
-) {
+// handleSubscription checks for a subscription within the given request and
+// starts a new go routine that will return all subscription results on the returned
+// channel. If a subscription does not exist on the given request nil will be returned.
+func (db *db) handleSubscription(ctx context.Context, r *request.Request) (<-chan client.GQLResult, error) {
 	if len(r.Subscription) == 0 || len(r.Subscription[0].Selections) == 0 {
-		// This is not a subscription request and we have nothing to do here
-		return nil, nil, nil
+		return nil, nil // This is not a subscription request and we have nothing to do here
 	}
-
 	if !db.events.Updates.HasValue() {
-		return nil, nil, ErrSubscriptionsNotAllowed
+		return nil, ErrSubscriptionsNotAllowed
 	}
-
-	s := r.Subscription[0].Selections[0]
-	if subRequest, ok := s.(*request.ObjectSubscription); ok {
-		pub, err := events.NewPublisher(db.events.Updates.Value(), 5)
-		if err != nil {
-			return nil, nil, err
-		}
-
-		return pub, subRequest, nil
+	selections := r.Subscription[0].Selections[0]
+	subRequest, ok := selections.(*request.ObjectSubscription)
+	if !ok {
+		return nil, client.NewErrUnexpectedType[request.ObjectSubscription]("SubscriptionSelection", selections)
+	}
+	// unsubscribing from this publisher will cause a race condition
+	// https://github.com/sourcenetwork/defradb/issues/2687
```

Comment:
question: Is this a new issue, or was the old code affected to?

And where is the race condition?  Is it a test artefact, or production code?

Reply:
I believe this is an old issue. The previous code never called `Unsubscribe`. The race happens between the `db.Close` and the `pub.Unsubscribe` so it's possible to happen in production, but likely would cause no issues.